docker-api/src/main/java/com/yahoo/vespa/hosted/dockerapi/DockerImageGarbageCollector.java

```diff
@@ -3,107 +3,171 @@
 
 import com.github.dockerjava.api.model.Image;
 import com.github.dockerjava.api.model.Container;
+import com.google.common.base.Strings;
+import com.yahoo.collections.Pair;
 
+import java.time.Clock;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.HashMap;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.logging.Logger;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
+ * This class keeps track of downloaded docker images and helps delete images that have not been recently used
+ *
+ * <p>Definitions:
+ * <ul>
+ *   <li>Every image has exactly 1 id</li>
+ *   <li>Every image has between 0..n tags</li>
```

Comment:
Added pointer to doc, in short, it is all of the above. It is even possible to, for example, pull an image with one name and then create a local tag for that image to something else.

Reply:
that the url is a tag doesn't follow from that doc? It says it CAN be. But the important thing for us is that it MUST be. Perhaps that's something Screwdriver ensures for us.

Reply:
Yes, it is CAN since there can also be no tag. However; because we send it the same tag in the except list as we use to pull that image (in the hosted task), we are guaranteed they are the same.

Reply:
Right, but why does it have the tag of the pull URL? Is that added at the pull, at image creation?
...
Yup, this is only guaranteed because we're using Screwdriver:
... ['/usr/bin/docker', 'build', '--no-cache=True', '--pull=true', '--file=Dockerfile', '--tag', 'docker-registry.ops.yahoo.com:4443/vespa/ci:6.289.11', u'.']
... Tagging docker-registry.ops.yahoo.com:4443/vespa/ci:6.289.11 as docker-registry.ops.yahoo.com:4443/vespa/ci:latest